packages/@aws-cdk/aws-codebuild/lib/project.ts

```diff
@@ -728,7 +728,7 @@ export class Project extends ProjectBase {
       const cfnEnvVariable: CfnProject.EnvironmentVariableProperty = {
         name,
         type: envVariable.type || BuildEnvironmentVariableType.PLAINTEXT,
-        value: envVariableValue,
+        value: envVariable.value instanceof Object ? (envVariable.value as secretsmanager.Secret).secretArn : envVariableValue,
```

Comment:
Yeah, this is a valiant effort, but definitely not the correct condition 🙂.

We need to use a TypeScript trick that we employ in a few other places as well. It looks something like this:

```ts
// this is in the SecretsManager module

const SECRET_SYMBOL = Symbol.for('@aws-cdk/core.Construct');

abstract class SecretBase {
  constructor(scope: Construct, id: string, props: ResourceProps = {}) {
    super(scope, id, props); 

    Object.defineProperty(this, SECRET_SYMBOL, { value: true });
  }

  // ...
}

export class Secret {
  // ...

  public static isSecret(x: any): x is ISecret {
    return x !== null && typeof(x) === 'object' && SECRET_SYMBOL in x;
  }

  // ...
}
```

And then, in CodeBuild, you can do `if (Secret.isSecret(envVariable.value)) ...`.

Reply:
I think we will have a problem when we want to support json keys. Since we only parse the Arn here for the value there is no way to supply the json key or am I missing something here? Seems like we would need to allow passing another string to enable json keys.

Any thoughts @skinny85 ?